### ISSUE TITLE
Handling signal events

### ISSUE THREAD

**Issue by MaheshkumarSundaram** [ralphhanna/bpmn-server#51]
_Using 1.3.4_

### Issue 1:

Consider the below example process:
![image](https://user-images.githubusercontent.com/55505611/188130196-7b1e6fbe-bc54-444f-b300-66c22fbb4254.png)

<details>
<summary>BPMN File</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="Definitions_1a44rne" targetNamespace="http://bpmn.io/schema/bpmn" exporter="bpmn-js (https://demo.bpmn.io)" exporterVersion="9.3.2">
  <bpmn:process id="Process_0r2ybht" isExecutable="false">
    <bpmn:startEvent id="workflow_start" name="Start">
      <bpmn:outgoing>Flow_1w74ars</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:sequenceFlow id="Flow_1w74ars" sourceRef="workflow_start" targetRef="loading_task" />
    <bpmn:userTask id="loading_task" name="Loading">
      <bpmn:incoming>Flow_1w74ars</bpmn:incoming>
      <bpmn:outgoing>Flow_1r48lhu</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:sequenceFlow id="Flow_1r48lhu" sourceRef="loading_task" targetRef="planning_task" />
    <bpmn:sequenceFlow id="Flow_0yn3lbi" sourceRef="planning_task" targetRef="auto_check_task" />
    <bpmn:serviceTask id="auto_check_task" name="AutoCheck" implementation="rtpAutoCheck" camunda:delegateExpression="rtpAutoCheck">
      <bpmn:extensionElements>
        <camunda:inputOutput>
          <camunda:outputParameter name="autoCheckStatus">output.autoCheckStatus</camunda:outputParameter>
        </camunda:inputOutput>
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_0yn3lbi</bpmn:incoming>
      <bpmn:outgoing>Flow_0rivcwh</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id="Flow_0rivcwh" sourceRef="auto_check_task" targetRef="auto_check_gateway" />
    <bpmn:userTask id="check_task" name="Checking">
      <bpmn:incoming>Flow_11cwkoj</bpmn:incoming>
      <bpmn:outgoing>Flow_0qwuw2x</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:exclusiveGateway id="check_gateway" name="Check Status of Checking" default="Flow_1yczic8">
      <bpmn:incoming>Flow_0qwuw2x</bpmn:incoming>
      <bpmn:outgoing>Flow_1b6byb4</bpmn:outgoing>
      <bpmn:outgoing>Flow_1yczic8</bpmn:outgoing>
    </bpmn:exclusiveGateway>
    <bpmn:sequenceFlow id="Flow_0qwuw2x" sourceRef="check_task" targetRef="check_gateway" />
    <bpmn:sequenceFlow id="Flow_1b6byb4" name="CheckPass" sourceRef="check_gateway" targetRef="generate_text_file">
      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression" language="JavaScript">(item.data.checkStatus === "Pass")</bpmn:conditionExpression>
    </bpmn:sequenceFlow>
    <bpmn:serviceTask id="generate_text_file" name="Generate Text File" implementation="rtpGenerateFile" camunda:delegateExpression="rtpGenerateFile">
      <bpmn:incoming>Flow_1b6byb4</bpmn:incoming>
      <bpmn:outgoing>Flow_0bz5tod</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:exclusiveGateway id="auto_check_gateway" name="Check Status of AutoCheck" default="Flow_11cwkoj">
      <bpmn:incoming>Flow_0rivcwh</bpmn:incoming>
      <bpmn:outgoing>Flow_11cwkoj</bpmn:outgoing>
      <bpmn:outgoing>Flow_0qwl78t</bpmn:outgoing>
    </bpmn:exclusiveGateway>
    <bpmn:sequenceFlow id="Flow_11cwkoj" name="AutoCheckPass" sourceRef="auto_check_gateway" targetRef="check_task" />
    <bpmn:sequenceFlow id="Flow_0qwl78t" name="AutoCheckFail" sourceRef="auto_check_gateway" targetRef="planning_task">
      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression" language="JavaScript">(item.data.autoCheckStatus === "Fail")</bpmn:conditionExpression>
    </bpmn:sequenceFlow>
    <bpmn:userTask id="planning_task" name="Planning">
      <bpmn:incoming>Flow_1r48lhu</bpmn:incoming>
      <bpmn:incoming>Flow_0qwl78t</bpmn:incoming>
      <bpmn:incoming>Flow_04tf59t</bpmn:incoming>
      <bpmn:outgoing>Flow_0yn3lbi</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:endEvent id="workflow_end" name="End">
      <bpmn:incoming>Flow_0bz5tod</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="Flow_0bz5tod" sourceRef="generate_text_file" targetRef="workflow_end" />
    <bpmn:sequenceFlow id="Flow_04tf59t" sourceRef="replan_catch" targetRef="planning_task" />
    <bpmn:intermediateCatchEvent id="replan_catch" name="Replan">
      <bpmn:outgoing>Flow_04tf59t</bpmn:outgoing>
      <bpmn:signalEventDefinition id="SignalEventDefinition_15rsrh1" signalRef="Signal_04pyjv4" />
    </bpmn:intermediateCatchEvent>
    <bpmn:sequenceFlow id="Flow_1yczic8" name="checkFail" sourceRef="check_gateway" targetRef="Event_059jpqk" />
    <bpmn:intermediateThrowEvent id="Event_059jpqk" name="Replan">
      <bpmn:incoming>Flow_1yczic8</bpmn:incoming>
      <bpmn:signalEventDefinition id="SignalEventDefinition_1jebn3n" signalRef="Signal_1mrpgls" />
    </bpmn:intermediateThrowEvent>
  </bpmn:process>
  <bpmn:signal id="Signal_1mrpgls" name="replan_signal" />
  <bpmn:signal id="Signal_04pyjv4" name="replan_signal" />
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_0r2ybht">
      <bpmndi:BPMNEdge id="Flow_1yczic8_di" bpmnElement="Flow_1yczic8">
        <di:waypoint x="1120" y="145" />
        <di:waypoint x="1120" y="232" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="1111" y="186" width="48" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_04tf59t_di" bpmnElement="Flow_04tf59t">
        <di:waypoint x="420" y="232" />
        <di:waypoint x="420" y="160" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0bz5tod_di" bpmnElement="Flow_0bz5tod">
        <di:waypoint x="1350" y="120" />
        <di:waypoint x="1462" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0qwl78t_di" bpmnElement="Flow_0qwl78t">
        <di:waypoint x="780" y="145" />
        <di:waypoint x="780" y="200" />
        <di:waypoint x="450" y="200" />
        <di:waypoint x="450" y="160" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="579" y="182" width="72" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_11cwkoj_di" bpmnElement="Flow_11cwkoj">
        <di:waypoint x="805" y="120" />
        <di:waypoint x="920" y="120" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="824" y="102" width="78" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1b6byb4_di" bpmnElement="Flow_1b6byb4">
        <di:waypoint x="1145" y="120" />
        <di:waypoint x="1250" y="120" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="1169" y="102" width="57" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0qwuw2x_di" bpmnElement="Flow_0qwuw2x">
        <di:waypoint x="1020" y="120" />
        <di:waypoint x="1095" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0rivcwh_di" bpmnElement="Flow_0rivcwh">
        <di:waypoint x="660" y="120" />
        <di:waypoint x="755" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0yn3lbi_di" bpmnElement="Flow_0yn3lbi">
        <di:waypoint x="500" y="120" />
        <di:waypoint x="560" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1r48lhu_di" bpmnElement="Flow_1r48lhu">
        <di:waypoint x="340" y="120" />
        <di:waypoint x="400" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1w74ars_di" bpmnElement="Flow_1w74ars">
        <di:waypoint x="188" y="120" />
        <di:waypoint x="240" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="workflow_start">
        <dc:Bounds x="152" y="102" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="158" y="145" width="25" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_15ono2o_di" bpmnElement="loading_task">
        <dc:Bounds x="240" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0gt15u4_di" bpmnElement="auto_check_task">
        <dc:Bounds x="560" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1951bm2_di" bpmnElement="check_task">
        <dc:Bounds x="920" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_0id7guv_di" bpmnElement="check_gateway" isMarkerVisible="true">
        <dc:Bounds x="1095" y="95" width="50" height="50" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="1084" y="65" width="79" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_06sqx9e_di" bpmnElement="generate_text_file">
        <dc:Bounds x="1250" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_1r8d4g2_di" bpmnElement="auto_check_gateway" isMarkerVisible="true">
        <dc:Bounds x="755" y="95" width="50" height="50" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="741" y="65" width="79" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_184uw60_di" bpmnElement="planning_task">
        <dc:Bounds x="400" y="80" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0wlem5q_di" bpmnElement="workflow_end">
        <dc:Bounds x="1462" y="102" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="1470" y="145" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_02xcwur_di" bpmnElement="replan_catch">
        <dc:Bounds x="402" y="232" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="403" y="275" width="35" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_1sjl09s_di" bpmnElement="Event_059jpqk">
        <dc:Bounds x="1102" y="232" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="1103" y="275" width="35" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
```
</details>

#### Expected Scenario:
When the intermediate signal event is thrown, the catch picks it up & starts the user task "Planning".

#### Actual Scenario:
The process stops execution once the intermediate signal is thrown.

If the signal is thrown from one process and caught in another, the execution is perfect. 
How to configure signal throw & catch in a single process?

### Issue 2:

This is in relation to the examples `test-throw` and `test-catch` processes from #46. There is nothing in their execution. Its as expected.

But the problem is the following scenario:

1. Say if there are two instances of `test-throw` which has been triggered at different times.
2. The first instance of `test-throw` throws a message event & triggers the first instance of `test-catch`
3. Likewise, the first instance of `test-throw` throws a message event & triggers the second instance of `test-catch`
4. Until now it's as expected
5. Next, when the first instance of `test-catch` sends a end signal event; it's caught by both the instances of `test-throw` and they resume.

I believe `messageMatchingKey` is used to handle similar scenarios in case of messages. 
Could you please let me know how to configure the **signal** events to target a specific instance?

Thank you!

**Comment by ralphhanna** [comment]
How to use the messageMatchKey
---------------------------------------
[Please see](https://github.com/ralphhanna/bpmn-server/blob/master/docs/examples.md#throwing-and-cathcing-messages)

**messageMatchKey** is set like setting output data at time of throwing the message and is used to filter the catching items.
MessageMatchKey is usually, caseId or clientId or any other value to restrict the scope of communications between instances and is assigned as part of data for both instances, in your case I believe you already set caseId to be same.

```js
this.messageMatchingKey={'data.caseId': data.caseId };
```
Normally, the system looks for items with messageName:
```js
findItems({'items.messageId': <messageName>,'items.status':'wait'});
```
But in this case it as the conditions of the messageMatchKey as follows:
```js
findItems({messageId: <messageName>,'items.status':'wait', 'data.caseId': <data.caseId> });
```

Same logic applies to Signals

**Comment by MaheshkumarSundaram** [comment]
Thanks for the explanation. Does the use of `matchingKey` solve the issue scenario 1 too?

**Comment by ralphhanna** [comment]
How to configure signal throw & catch in a single process?
----------------------------------------------------------------
Signals and Messages are not intended to be used for same process, however, you can perform this using different processes in same definition see example of "Invoice".

** But why not use simple flow instead of signals? **

![image](https://user-images.githubusercontent.com/11893416/188203226-b30bc537-1ffa-4b29-bc36-ee3ad8f5eed2.png)

Much simpler!

**Comment by MaheshkumarSundaram** [comment]
Tbh, I was using the gateway flows like you mentioned.

I just wanted to try whether signal could be used this way too. 
Now that you have mentioned that both Signals & Messages are used to coordinate different processes. That clears the air.

Thank you!

**Comment by MaheshkumarSundaram** [comment]
The reason I wanted to try the signal events in a process is for some complex processes which I have.

**For example:**
![image](https://user-images.githubusercontent.com/55505611/188206298-c6626b5f-d555-476f-806c-6a4b5aa18d51.png)
_I couldn't even capture the full process 😆._ 

As you can see in the above diagram, there are so many gateways. Modelling flows for all those gateways could be like a cluttering. So I thought I could leverage the use of signal & message events.

Now that the messages and signals can't be thrown & caught inside the same process, I guess I will remodel my process and split them into separate ones logically.

If you have any suggestions regarding this, kindly let me know @ralphhanna 
Thanks!

**Comment by ralphhanna** [comment]
Issue #1 
---------

![image](https://user-images.githubusercontent.com/11893416/188211349-54222a03-41fe-4856-b4dd-f874e388bf17.png)

The process stops because it has no where to go? If you want with this design, you can do this: 
![image](https://user-images.githubusercontent.com/11893416/188211836-2b75561d-3446-4170-8a24-6e4409db24ee.png)

But wouldn't recommend it.

Thanks

**Comment by ralphhanna** [comment]
Found some more issues with signals.
Fixed in release 1.3.5
Please update.

**Comment by MaheshkumarSundaram** [comment]
Okay. Thanks!
